components/dashboard/src/settings/Integrations.tsx

```diff
@@ -199,77 +179,435 @@ function GitProviders() {
+            </div>
+        )}
+        <div className="flex flex-col pt-6 space-y-2">
+            {providers && providers.map(ap => (
+                <div key={"ap-" + ap.id} className="flex-grow flex flex-row hover:bg-gray-100 rounded-xl h-16 w-full">
+
+                    <div className="px-4 self-center w-1/12">
+                        <div className={"rounded-full w-3 h-3 text-sm align-middle " + (ap.status === "verified" ? "bg-green-500" : "bg-gray-400")}>
+                            &nbsp;
+                        </div>
+                    </div>
+                    <div className="p-0 my-auto flex flex-col w-3/12">
+                        <span className="my-auto font-medium truncate overflow-ellipsis">{ap.type}</span>
                     </div>
-                    <div className="self-center">
-                        <ContextMenu menuEntries={gitProviderMenu(ap)}>
-                            <img className="w-8 h-8 p-1" src={ThreeDots} alt="Actions" />
-                        </ContextMenu>
+                    <div className="p-0 my-auto flex flex-col w-7/12">
+                        <span className="my-auto truncate text-gray-500 overflow-ellipsis">{ap.host}</span>
+                    </div>
+                    <div className="my-auto flex w-1/12">
+                        <div className="self-center hover:bg-gray-200 rounded-md cursor-pointer w-8">
+                            <ContextMenu menuEntries={gitProviderMenu(ap)}>
+                                <img className="w-8 h-8 p-1" src={ThreeDots} alt="Actions" />
+                            </ContextMenu>
+                        </div>
                     </div>
                 </div>
             ))}
         </div>
+        {providers && providers.length > 0 && (
+            <div className="flex justify-start mt-6">
+                <button className="h-full" onClick={() => setModal({ mode: "new" })}>New Git Integration</button>
+            </div>
+        )}
     </div>);
 }
 
+function GitIntegrationModal(props: ({
+    mode: "new",
+} | {
+    mode: "edit",
+    provider: AuthProviderEntry
+}) & {
+    userId: string,
+    onClose?: () => void
+    update?: () => void
+}) {
+
+    const callbackUrl = (host: string) => {
+        const pathname = `/auth/${host}/callback`;
+        return gitpodHostUrl.with({ pathname }).toString();
+    }
+
+    const [type, setType] = useState<string>("GitLab");
+    const [host, setHost] = useState<string>("gitlab.example.com");
+    const [redirectURL, setRedirectURL] = useState<string>(callbackUrl("gitlab.example.com"));
+    const [clientId, setClientId] = useState<string>("");
+    const [clientSecret, setClientSecret] = useState<string>("");
+    const [busy, setBusy] = useState<boolean>(false);
+    const [errorMessage, setErrorMessage] = useState<string | undefined>();
+    const [validationError, setValidationError] = useState<string | undefined>();
+
+    const firstRender = useRef(true);
```

Comment:
Tracking how often it is rendered looks like you tried to fix something weird. 
What's the purpose of `firstRender`? If you think this is the right solution, please at least explain in a comment.

Reply:
why so? 
the first render thingy seems to be a common case, https://stackoverflow.com/questions/53253940/make-react-useeffect-hook-not-run-on-initial-render

the effect here is, that I don't want the validation on `[clientId, clientSecret]` being triggered right away, but only on user changes.